### PR TITLE
Sampling Performance Testing

### LINKED ISSUE #3839: Add Reference CuGraphSAGE Implementation to cuGraph-PyG
Just as the native frameworks have their default models, we should offer the same.  This will consolidate code across all of our examples, and provide a clear example for users on how to build a cuGraph-PyG/cuGraph-ops model.  This issue came about when working on slides for the PyG conference and needing a central model example to point users to.

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
Adds performance benchmarking scripts for testing MNMG cuGraph GNN workflows.
This branch is the head branch for the cuGraph benchmarking effort.  All work supporting the benchmarks should be merged into this branch.  It will be merged into branch-24.02 once all features are ready.

Includes patches to cuGraph-PyG required for the latest DLFW container.

To-Do:

- [x] Refactor for branch-24.02
- [x] <s>Add WholeGraph training portion</s> Deferred to future PR (see https://github.com/alexbarghi-nv/cugraph/pull/6)
- [x] <s>Add WholeGraph generators</s> Included in above
- [x] <s>Support DGL</s> Deferred to future PR
- [x] <s>Use appropriate docker containers</s> Deferred, waiting on DLFW release

Closes #3839

### CODE REVIEW

File: benchmarks/cugraph/standalone/cugraph_bulk_sampling.py [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+    return (end_time - start_time)
+
+
+def load_disk_dataset(dataset, dataset_dir='.', reverse_edges=True, replication_factor=1):
+    path = os.path.join(dataset_dir, dataset)
+    parquet_path = os.path.join(path, 'parquet')
+
+    with open(os.path.join(path, 'meta.json')) as meta_file:
+        meta = json.load(meta_file)
+
```
Comment (VibhuJawa) [inline benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:195]: Can i have an example of `meta.json`. 

I basically need below for DGL dataloaders. https://github.com/rapidsai/cugraph/blob/c49643c1263bb5c0a00a5f9c2c0b995a075a1396/python/cugraph-dgl/cugraph_dgl/dataloading/dataset.py#L83-L87
```python3
num_nodes_dict: Dict[str, int],
etype_id_dict: Dict[int, Tuple[str, str, str]],
etype_offset_dict: Dict[Tuple[str, str, str], int],
ntype_offset_dict: Dict[str, int],
```

Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:195]: I can add offsets to the `output_meta.json` if that will make it easier for DGL

Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:195]: Output meta example:
```
{
"dataset": "ogbn_papers100M",
"dataset_dir": "/datasets/abarghi",
"seed": 62,
"batch_size": 500,
"seeds_per_call": 500000,
"fanout": [10, 25],
"replication_factor": 2,
"execution_time": 10.963991190772504,
"offsets": {
     "paper": 0
}
}
```

Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:195]: In that script I put a comment that it's using the cuGraph-PyG offset conventions

Comment (VibhuJawa) [reply benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:195]: For homogenous graphs, I dont need anything but `total_number_of_nodes `


https://github.com/rapidsai/cugraph/blob/c49643c1263bb5c0a00a5f9c2c0b995a075a1396/python/cugraph-dgl/cugraph_dgl/dataloading/dataset.py#L33-L34

Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:195]: Which my loader script uses for heterogeneous graphs

Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:195]: I can output those offsets in the output_meta.json file so that you can use them.

Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:195]: ok, I'll work on that

File: benchmarks/cugraph/standalone/cugraph_bulk_sampling.py [hunk 2]
```diff
@@ -0,0 +1,617 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
```
Comment (VibhuJawa) [inline benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:1]: I think we can probably separate this file and merge it as separate PR ?

Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/cugraph_bulk_sampling.py:1]: Yes, I'll take care of that today.

File: benchmarks/cugraph-pyg/cugraph_pyg_graph_sage.py [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+            print(f'Loading edge index for edge type {edge_type}')
+
+            can_edge_type = tuple(edge_type.split('__'))
+            ei = pandas.read_parquet(os.path.join(os.path.join(parquet_path, edge_type), 'edge_index.parquet'))
+            ei = {
+                'src': torch.from_numpy(ei.src.values),
+                'dst': torch.from_numpy(ei.dst.values),
+            }
+            print('sorting edge index...')
+            ei['dst'], ix = torch.sort(ei['dst'])
```
Comment (VibhuJawa) [inline benchmarks/cugraph-pyg/cugraph_pyg_graph_sage.py:213]: Curios on Why are we sorting here

File: benchmarks/cugraph-pyg/cugraph_pyg_graph_sage.py [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+
+    # Have to load graph data for native PyG
+    parquet_path = os.path.join(
+        output_meta['dataset_dir'],
+        output_meta['dataset'],
+        'parquet'
+    )
+
+    for edge_type in os.listdir(parquet_path):
+        if re.match(r'[a-z]+__[a-z]+__[a-z]+', edge_type):
```
Comment (VibhuJawa) [inline benchmarks/cugraph-pyg/cugraph_pyg_graph_sage.py:203]: I suggest using `meta.json` to iterate through edges as that is more concrete

File: benchmarks/cugraph-pyg/cugraph_pyg_graph_sage.py (hunk 3 above)
Comment (alexbarghi-nv) [reply benchmarks/cugraph-pyg/cugraph_pyg_graph_sage.py:213]: Need to sort before converting to CSC

File: benchmarks/cugraph-pyg/cugraph_pyg_graph_sage.py (hunk 4 above)
Comment (alexbarghi-nv) [reply benchmarks/cugraph-pyg/cugraph_pyg_graph_sage.py:203]: I'll update that here, too

File: benchmarks/cugraph/standalone/bulk_sampling/run_train_job.sh [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#SBATCH -A datascience_rapids_cugraphgnn
+#SBATCH -p luna
+#SBATCH -J datascience_rapids_cugraphgnn-papers:bulkSamplingPyG
+#SBATCH -N 1
+#SBATCH -t 00:25:00 
+
+CONTAINER_IMAGE="/lustre/fsw/rapids/abarghi/dlfw_patched.squash"
```
Comment (alexbarghi-nv) [inline benchmarks/cugraph/standalone/bulk_sampling/run_train_job.sh:21]: Will replace with correct DLFW image once it is available

File: benchmarks/cugraph/standalone/bulk_sampling/README.md [hunk 6]
```diff
@@ -103,14 +99,64 @@ the parquet files.  It must have the following format:
...
+Next are arguments for the container image (required),
+and directories where the data and outputs are stored.  The directories default to subdirectories
+of the current working directory.  But if there is a high-throughput storage system available,
+using that storage for the samples and datasets is highly recommended.
+
+Next are standard GNN training arguments such as `FANOUT`, `BATCH_SIZE`, etc.  You can also set
+the number of training epochs here.  These are followed by the `REPLICATION_FACTOR` argument, which
+can be used to create replications of the dataset for scale testing purposes.
+
+The final two arguments are `FRAMEWORK` which can be either "cuGraphPyG" or "PyG", and `GPUS_PER_NODE`
```
Comment (VibhuJawa) [inline benchmarks/cugraph/standalone/bulk_sampling/README.md:155]: I assume we shall include "cuGraphDGL" here too.

File: benchmarks/cugraph/standalone/bulk_sampling/bench_cugraph_training.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+os.environ["RAPIDS_NO_INITIALIZE"] = "1"
+os.environ["CUDF_SPILL"] = "1"
+os.environ["LIBCUDF_CUFILE_POLICY"] = "KVIKIO"
+os.environ["KVIKIO_NTHREADS"] = "64"
```
Comment (VibhuJawa) [inline benchmarks/cugraph/standalone/bulk_sampling/bench_cugraph_training.py:19]: We should probably reduce to "8" , as if we are using multiple GPUs , we get 8 per GPU so 64 in total.

File: benchmarks/cugraph/standalone/bulk_sampling/run_sampling.sh [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+source ~/.bashrc
+conda activate rapids
+
+BATCH_SIZE=$1
+FANOUT=$2
+REPLICATION_FACTOR=$3
+SCRIPTS_DIR=$4
+NUM_EPOCHS=$5
+
+SAMPLES_DIR=/samples
```
Comment (VibhuJawa) [inline benchmarks/cugraph/standalone/bulk_sampling/run_sampling.sh:25]: Assuming we are mounting this to the most performant path.

File: mg_utils/wait_for_workers.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
+# Copyright (c) 2023-2024, NVIDIA CORPORATION.
```
Comment (VibhuJawa) [inline mg_utils/wait_for_workers.py:1]: I dont want to break our current workflow but as a non blocking item can we file an issue for investigating this vs  `client.wait_for_workers` .  Lets file an issue and add a note here. 

```
https://distributed.dask.org/en/stable/api.html#distributed.Client.wait_for_workers
```

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Requested minor changes, mostly looks good.

File: benchmarks/cugraph/standalone/bulk_sampling/README.md (hunk 6 above)
Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/bulk_sampling/README.md:155]: Yes

Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/bulk_sampling/README.md:155]: in the next PR

File: benchmarks/cugraph/standalone/bulk_sampling/run_sampling.sh (hunk 8 above)
Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/bulk_sampling/run_sampling.sh:25]: Yes, it is up to us to set `LOGS_DIR`, `SAMPLES_DIR`, and `DATASETS_DIR` in `run_train_job.sh` correctly.  In the `srun` command, those are mounted to `/logs`, `/samples`, and `/datasets` in the container that this script runs in.

File: mg_utils/wait_for_workers.py (hunk 9 above)
Comment (alexbarghi-nv) [reply mg_utils/wait_for_workers.py:1]: https://github.com/rapidsai/cugraph/issues/4082

File: benchmarks/cugraph/standalone/bulk_sampling/bench_cugraph_training.py (hunk 7 above)
Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/bulk_sampling/bench_cugraph_training.py]: fixed

**Review by VibhuJawa — APPROVED** [review]
LGTM

**Review by rlratzel — APPROVED** [review]
LGTM overall.  I don't feel too strongly, but I noticed several places in the shell scripts that assume things about the file system (`/datasets`, etc.). We usually put those scripts in another repo (the repo containing our machine-specific nightly scripts, etc.) and not the open cugraph repo.

**Comment by alexbarghi-nv** [comment]
> LGTM overall. I don't feel too strongly, but I noticed several places in the shell scripts that assume things about the file system (`/datasets`, etc.). We usually put those scripts in another repo (the repo containing our machine-specific nightly scripts, etc.) and not the open cugraph repo.

The `/datasets` and `/scripts` directories are mounted to directories the user has to provide.  They default to the current working directory (it creates folders there and mounts them).  So we're not exposing any NVIDIA internal info.

**Pull request merged**